File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/first/FirstScreen.java
package com.jaynewstrom.screenswitchersample.first;

import android.content.Context;
import android.view.View;

import com.jaynewstrom.concrete.ConcreteBlock;
import com.jaynewstrom.screenswitcher.ScreenTransition;
import com.jaynewstrom.screenswitchersample.base.DefaultScreenTransition;
import com.jaynewstrom.screenswitchersample.base.ParentComponent;
import com.jaynewstrom.screenswitchersample.base.concrete.ConcreteScreen;

public final class FirstScreen extends ConcreteScreen<FirstComponent> {

    @Override public ScreenTransition transition() {
        return DefaultScreenTransition.INSTANCE;
    }

    @Override protected ConcreteBlock<FirstComponent> block(ParentComponent parentComponent) {
        return new FirstScreenBlock(parentComponent);
    }

    @Override public View createView(Context context, FirstComponent component) {
        return new FirstView(context, component);
    }
}

File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/first/FirstScreen.kt
package com.jaynewstrom.screenswitchersample.first

import android.content.Context
import android.view.View
import android.view.ViewGroup
import com.jaynewstrom.concrete.ConcreteBlock
import com.jaynewstrom.screenswitcher.Screen
import com.jaynewstrom.screenswitchersample.core.BaseScreen
import com.jaynewstrom.screenswitchersample.core.DefaultScreenWallManager
import com.jaynewstrom.screenswitchersample.core.ScreenScope
import com.jaynewstrom.screenswitchersample.core.ScreenWallManager
import dagger.Component

object FirstScreenFactory {
    fun create(): Screen = FirstScreen()
}

private class FirstScreen : BaseScreen<FirstComponent>() {
    override fun createWallManager(): ScreenWallManager<FirstComponent> {
        return DefaultScreenWallManager({
            FirstScreenBlock()
        })
    }

    public override fun createView(context: Context, hostView: ViewGroup, component: FirstComponent): View {
        return FirstPresenter.createView(context, hostView, component)
    }
}

@ScreenScope
@Component
internal interface FirstComponent {
    fun inject(presenter: FirstPresenter)
}

private class FirstScreenBlock : ConcreteBlock<FirstComponent> {
    override fun name(): String = javaClass.name

    override fun createComponent(): FirstComponent = DaggerFirstComponent.create()
}
